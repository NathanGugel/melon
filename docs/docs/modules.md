---
id: modules
title: "Modules"
sidebar_label: "Modules"
slug: /modules
---

Melon has a basic but flexible module system. Modules are used by the core system to expose domain specific functionality to the developer and can also be used by the developer to enhance code re-use.

## Core modules

Core modules are exposed through global variables and are pre-defined. One example is the [`io` module](io_module.md) which was used throughout this documentation.

## Importing custom modules

Custom modules must be selectively imported by the developer, to do this `import` can be called:

```js
let MyModule = import("my-module");

MyModule.doSomething();
```

:::note
As will become apparent in the sections below, `import` is not a `Function` but an `Object`.
It can be called because it has its [`call operator`](operators.md#supported-custom-operators) overriden, allowing the developer to customise
its behaviour.
:::

### Import path

The first argument of the `import` call is used to locate the module's source. How this is done depends on the content of the `import.paths` key which contains an `Array` of strings:

```js
import.paths = [
    "##/??/index.ms",
    "##/../??/index.ms"
];
```

The path strings are processed top to bottom until a valid file is found.
Special placeholders can be used inside path strings:

| Placeholder | Meaning |
| ----------- | ------- |
| `??` | The first argument of the `import` call |
| `##` | If the current call to `import` has been made from `a/b/c.ms` then `##` is `a/b` |
| `$$` | The path of the entry-point. For example if your entry point is `a/b/main.ms` then `$$` is `a/b` |

The default value for `import.paths` is:

```
##/??
##/??.ms
##/??.mbc
##/??/index.ms
##/??/index.mbc
$$/??
$$/??.ms
$$/??.mbc
$$/??/index.ms
$$/??/index.mbc
$$/mel_seeds/??.ms
$$/mel_seeds/??.mbc
$$/mel_seeds/??/index.ms
$$/mel_seeds/??/index.mbc
$$/mel_seeds/??
??
```

:::note
`import` can also import pre-compiled bytecode, that's what the `mbc` extension stands for.
:::

## Defining custom modules

Custom modules are just standard source files which `return` a value. The value will be the result of the `import` call when the module is imported:

In `a/my-module.ms`:
```js
return {
    myVar = 5,
    myFunc = => { io.print("Hello modules!"); }
};
```

In `a/some-file.ms`:
```js
let MyModule = import("my-module");

io.print(MyModule.myVar); // 5
MyModule.myFunc(); // Hello modules!
```

### Caching

Modules are cached, this means the the module's code is executed only upon first `import`. Then the returned value is cached and the same value is returned to every subsequent `import` call. This means that if your module's code has any side-effects they will be executed only once:

In `a/my-module.ms`:
```js
io.print("Importing!");

return {
    myVar = 5,
    myFunc = => { io.print("Hello modules!"); }
};
```

In `a/some-file.ms`:
```js
let MyModule = import("my-module");
let MyModule2 = import("my-module");

MyModule.myFunc(); // Hello modules!
```

Running `some-file.ms` will output:
```
Importing!
Hello modules!
```

### Accessing the cache

`import`'s cache is exposed through its `import.cache` key which is just an `Object` with the key corresponding to the real path on the filesystem of the imported file and the value to the value returned by the module. You can for example clear all `import`'s cache by simply doing:

```js
import.cache = {};
```