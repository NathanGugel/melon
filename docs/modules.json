["array_module","cli_module","compiler_module","debug_module","fs_module","function_module","gc_module","import_module","integer_module","io_module","json_module","math_module","number_module","object_module","path_module","random_module","string_module","system_module","types_module"]