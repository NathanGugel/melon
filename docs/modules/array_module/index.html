<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-alpha.66">
<title data-react-helmet="true">array | melon</title><meta data-react-helmet="true" name="twitter:card" content="summary_large_image"><meta data-react-helmet="true" name="docusaurus_language" content="en"><meta data-react-helmet="true" name="docusaurus_version" content="current"><meta data-react-helmet="true" name="docusaurus_tag" content="docs-default-current"><meta data-react-helmet="true" property="og:title" content="array | melon"><meta data-react-helmet="true" name="description" content="This module defines utility functions for Array values manipulation."><meta data-react-helmet="true" property="og:description" content="This module defines utility functions for Array values manipulation."><meta data-react-helmet="true" property="og:url" content="https://daniele-rapagnani.github.io/melon/docs/modules/array_module"><link data-react-helmet="true" rel="shortcut icon" href="/melon/img/favicon.png"><link data-react-helmet="true" rel="canonical" href="https://daniele-rapagnani.github.io/melon/docs/modules/array_module"><link rel="stylesheet" href="/melon/styles.6874ab0a.css">
<link rel="preload" href="/melon/styles.bfc3e8ea.js" as="script">
<link rel="preload" href="/melon/runtime~main.08f41170.js" as="script">
<link rel="preload" href="/melon/main.506e7c1e.js" as="script">
<link rel="preload" href="/melon/1.343061b0.js" as="script">
<link rel="preload" href="/melon/2.330af622.js" as="script">
<link rel="preload" href="/melon/43.f3832d12.js" as="script">
<link rel="preload" href="/melon/44.4b500b20.js" as="script">
<link rel="preload" href="/melon/935f2afb.fe183059.js" as="script">
<link rel="preload" href="/melon/17896441.45e4c31b.js" as="script">
<link rel="preload" href="/melon/f0865d0b.9f0373b5.js" as="script">
</head>
<body>
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><div aria-label="Navigation bar toggle" class="navbar__toggle" role="button" tabindex="0"><svg xmlns="http://www.w3.org/2000/svg" width="30" height="30" viewBox="0 0 30 30" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></div><a class="navbar__brand" href="/melon/"><img class="navbar__logo" src="/melon/img/melon_logo_symbol.png" alt="Melon Programming Language"><strong class="navbar__title">melon</strong></a><a aria-current="page" class="navbar__item navbar__link navbar__link--active" href="/melon/docs/">Docs</a></div><div class="navbar__items navbar__items--right"><div class="react-toggle react-toggle--disabled displayOnlyInLargeViewport_2aTZ"><div class="react-toggle-track"><div class="react-toggle-track-check"><span class="toggle_BsTx">ðŸŒœ</span></div><div class="react-toggle-track-x"><span class="toggle_BsTx">ðŸŒž</span></div></div><div class="react-toggle-thumb"></div><input type="checkbox" disabled="" aria-label="Dark mode toggle" class="react-toggle-screenreader-only"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div><div class="navbar-sidebar"><div class="navbar-sidebar__brand"><a class="navbar__brand" href="/melon/"><img class="navbar__logo" src="/melon/img/melon_logo_symbol.png" alt="Melon Programming Language"><strong class="navbar__title">melon</strong></a></div><div class="navbar-sidebar__items"><div class="menu"><ul class="menu__list"><li class="menu__list-item"><a aria-current="page" class="menu__link navbar__link--active" href="/melon/docs/">Docs</a></li></ul></div></div></div></nav><div class="main-wrapper"><div class="docPage_2gpo"><div class="docSidebarContainer_3_JD" role="complementary"><div class="sidebar_2urC"><div class="menu menu--responsive menu_5FrY"><button aria-label="Open Menu" aria-haspopup="true" class="button button--secondary button--sm menu__button" type="button"><svg aria-label="Menu" class="sidebarMenuIcon_Dm3K" xmlns="http://www.w3.org/2000/svg" height="24" width="24" viewBox="0 0 32 32" role="img" focusable="false"><title>Menu</title><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><ul class="menu__list"><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!">Melon</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/">Introduction</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Language</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/variables">Variables</a></li><li class="menu__list-item menu__list-item--collapsed"><a class="menu__link menu__link--sublist" href="#!" tabindex="0">Value Types</a><ul class="menu__list"><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/integer">Integer</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/number">Number</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/bool">Boolean</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/string">String</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/array">Array</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/object">Object</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/function">Function</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/range">Range</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/symbol">Symbol</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/iterator">Iterator</a></li><li class="menu__list-item"><a class="menu__link" tabindex="-1" href="/melon/docs/types/null">Null</a></li></ul></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/control-structures">Control structures</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/operators">Operators</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules">Modules</a></li><li class="menu__list-item"><a class="menu__link menu__link--sublist menu__link--active" href="#!" tabindex="0">Core modules</a><ul class="menu__list"><li class="menu__list-item"><a aria-current="page" class="menu__link menu__link--active active" tabindex="0" href="/melon/docs/modules/array_module">array</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/cli_module">cli</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/compiler_module">compiler</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/debug_module">debug</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/fs_module">fs</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/function_module">function</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/gc_module">gc</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/import_module">import</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/integer_module">integer</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/io_module">io</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/json_module">json</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/math_module">math</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/number_module">number</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/object_module">object</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/path_module">path</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/random_module">random</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/string_module">string</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/system_module">system</a></li><li class="menu__list-item"><a class="menu__link" tabindex="0" href="/melon/docs/modules/types_module">types</a></li></ul></li></ul></li></ul></li></ul></div></div></div><main class="docMainContainer_3EyW"><div class="container padding-vert--lg docItemWrapper_1EkI"><div class="row"><div class="col docItemCol_2ASc"><div class="docItemContainer_3QWW"><article><header><h1 class="docTitle_1Lrw">array</h1></header><div class="markdown"><p>This module defines utility functions for <a href="/melon/docs/types/array"><code>Array</code></a> values manipulation.</p><h2><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="functions"></a>Functions<a aria-hidden="true" tabindex="-1" class="hash-link" href="#functions" title="Direct link to heading">#</a></h2><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="resizearr-size"></a>resize(arr, size)<a aria-hidden="true" tabindex="-1" class="hash-link" href="#resizearr-size" title="Direct link to heading">#</a></h3><p>Ensures the size of an array is at least the one provided.  </p><table><thead><tr><th>Argument</th><th>Description</th></tr></thead><tbody><tr><td><strong>arr</strong></td><td>The array to resize</td></tr><tr><td><strong>size</strong></td><td>An integer with the new size</td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="lookuparr-index"></a>lookup(arr, index)<a aria-hidden="true" tabindex="-1" class="hash-link" href="#lookuparr-index" title="Direct link to heading">#</a></h3><p>Returns the value stored at a given index for a given array without rising any error if the index does not exist.   @return The requeste value if present, <code>null</code> otherwise</p><table><thead><tr><th>Argument</th><th>Description</th></tr></thead><tbody><tr><td><strong>arr</strong></td><td>The array</td></tr><tr><td><strong>index</strong></td><td>The index to lookup</td></tr></tbody></table><h3><a aria-hidden="true" tabindex="-1" class="anchor enhancedAnchor_2cZh" id="deletearr-start-end"></a>delete(arr, start, end)<a aria-hidden="true" tabindex="-1" class="hash-link" href="#deletearr-start-end" title="Direct link to heading">#</a></h3><p>Removes one or more elements from an array, resizing it if necessary.    </p><table><thead><tr><th>Argument</th><th>Description</th></tr></thead><tbody><tr><td><strong>arr</strong></td><td>The array</td></tr><tr><td><strong>start</strong></td><td>The index of the first element to remove</td></tr><tr><td><strong>end</strong></td><td>The index of the last element to remove</td></tr></tbody></table><p><strong>Returns:</strong> <code>true</code> on success, <code>false</code> otherwise</p></div></article><div class="margin-vert--xl"><div class="row"><div class="col"><a href="https://github.com/daniele-rapagnani/edit/master/docs/docs/array_module.md" target="_blank" rel="noreferrer noopener"><svg fill="currentColor" height="1.2em" width="1.2em" preserveAspectRatio="xMidYMid meet" viewBox="0 0 40 40" style="margin-right:0.3em;vertical-align:sub"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div></div></div><div class="margin-vert--lg"><nav class="pagination-nav" aria-label="Blog list page navigation"><div class="pagination-nav__item"><a class="pagination-nav__link" href="/melon/docs/modules"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Â« Modules</div></a></div><div class="pagination-nav__item pagination-nav__item--next"><a class="pagination-nav__link" href="/melon/docs/modules/cli_module"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">cli Â»</div></a></div></nav></div></div></div><div class="col col--3"><div class="tableOfContents_3SO_"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#functions" class="table-of-contents__link">Functions</a><ul><li><a href="#resizearr-size" class="table-of-contents__link">resize(arr, size)</a></li><li><a href="#lookuparr-index" class="table-of-contents__link">lookup(arr, index)</a></li><li><a href="#deletearr-start-end" class="table-of-contents__link">delete(arr, start, end)</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container"><div class="row footer__links"><div class="col footer__col"><h4 class="footer__title">Community</h4><ul class="footer__items"><li class="footer__item"><a href="https://github.com/daniele-rapagnani/melon" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub</a></li><li class="footer__item"><a href="https://stackoverflow.com/questions/tagged/melon" target="_blank" rel="noopener noreferrer" class="footer__link-item">Stack Overflow</a></li></ul></div></div><div class="text--center"><div>Copyright Â© 2021 Daniele Rapagnani.</div></div></div></footer></div>
<script src="/melon/styles.bfc3e8ea.js"></script>
<script src="/melon/runtime~main.08f41170.js"></script>
<script src="/melon/main.506e7c1e.js"></script>
<script src="/melon/1.343061b0.js"></script>
<script src="/melon/2.330af622.js"></script>
<script src="/melon/43.f3832d12.js"></script>
<script src="/melon/44.4b500b20.js"></script>
<script src="/melon/935f2afb.fe183059.js"></script>
<script src="/melon/17896441.45e4c31b.js"></script>
<script src="/melon/f0865d0b.9f0373b5.js"></script>
</body>
</html>