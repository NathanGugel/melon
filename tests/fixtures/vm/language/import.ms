for (let p in import.paths) {
    io.print(p);
}

let obj = import("modules/test_module");
obj.doStuff();

let obj2 = import("modules/test_module");
obj2.doStuff();

import("modules/side_effects");
import("modules/side_effects");

import.cache = {};

import("modules/side_effects");

let moduleTest = import("modules/module_obj");
moduleTest->printInfo();

io.print("Dumping modules cache:");

for (let c in import.cache) {
    io.print(c);
    io.print(import.cache[c]);
}

import.paths = [
    "##/modules/??.mbc",
    "##/??/index.ms"
];

import.cache = {};

moduleTest = import("module_obj");
moduleTest->printInfo();

let oldImport = import[object.symbols.callOperator];

import[object.symbols.callOperator] = |path| => {
    io.print("Importing: " .. path);
    return oldImport(path);
};

import.paths = [ "##/??.ms" ];

import("modules/test_module");
import("modules/test_module");