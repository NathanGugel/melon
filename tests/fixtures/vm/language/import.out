##/??
##/??.ms
##/??.mbc
##/??/index.ms
##/??/index.mbc
$$/mel_seeds/??.ms
$$/mel_seeds/??.mbc
$$/mel_seeds/??/index.ms
$$/mel_seeds/??/index.mbc
$$/mel_seeds/??
??
Should only print once!
Cool Module!
Cool Module!
This is a side-effect
This is a side-effect
fixtures/vm/language/modules/module_obj.ms
Submodule found
Dumping modules cache:
fixtures/vm/language/modules/submodule/index.ms
[Closure @anonymous@]
fixtures/vm/language/modules/side_effects.ms
null
fixtures/vm/language/modules/module_obj.ms
[Object]
fixtures/vm/language/modules/module_obj.mbc
Submodule found
Importing: modules/test_module
Should only print once!
Importing: modules/test_module
